tools/xybmsssim.cc

```diff
@@ -0,0 +1,286 @@
+  for (size_t c = 0; c < 3; ++c) {
+    for (size_t y = 0; y < a.ysize(); ++y) {
+      const float* JXL_RESTRICT in1 = a.PlaneRow(c, y);
+      const float* JXL_RESTRICT in2 = b.PlaneRow(c, y);
+      float* JXL_RESTRICT out = mul->PlaneRow(c, y);
+      for (size_t x = 0; x < a.xsize(); ++x) {
+        out[x] = in1[x] * in2[x];
+      }
+    }
+  }
+}
+
+// Temporary storage for Gaussian blur, reused for multiple images.
+class Blur {
+ public:
+  Blur(const size_t xsize, const size_t ysize)
+      : rg_(jxl::CreateRecursiveGaussian(1.5)), temp_(xsize, ysize) {}
+
+  void operator()(const ImageF& in, ImageF* JXL_RESTRICT out) {
+    jxl::ThreadPool* null_pool = nullptr;
+    FastGaussian(rg_, in, null_pool, &temp_, out);
+  }
+
+  Image3F operator()(const Image3F& in) {
+    Image3F out(in.xsize(), in.ysize());
+    operator()(in.Plane(0), &out.Plane(0));
+    operator()(in.Plane(1), &out.Plane(1));
+    operator()(in.Plane(2), &out.Plane(2));
+    return out;
+  }
+
+  // Allows reusing across scales.
+  void ShrinkTo(const size_t xsize, const size_t ysize) {
+    temp_.ShrinkTo(xsize, ysize);
+  }
+
+ private:
+  hwy::AlignedUniquePtr<jxl::RecursiveGaussian> rg_;
+  ImageF temp_;
+};
+
+void SSIMMap(const Image3F& m1, const Image3F& m2, const Image3F& s11,
+             const Image3F& s22, const Image3F& s12, Image3F* out,
+             double* plane_averages) {
+  const double onePerPixels = 1.0 / (out->ysize() * out->xsize());
+  for (size_t c = 0; c < 3; ++c) {
+    double sum1[4] = {0.0};
+    for (size_t y = 0; y < out->ysize(); ++y) {
+      const float* JXL_RESTRICT row_m1 = m1.PlaneRow(c, y);
+      const float* JXL_RESTRICT row_m2 = m2.PlaneRow(c, y);
+      const float* JXL_RESTRICT row_s11 = s11.PlaneRow(c, y);
+      const float* JXL_RESTRICT row_s22 = s22.PlaneRow(c, y);
+      const float* JXL_RESTRICT row_s12 = s12.PlaneRow(c, y);
+      float* JXL_RESTRICT row_out = out->PlaneRow(c, y);
+      for (size_t x = 0; x < out->xsize(); ++x) {
+        float mu1 = row_m1[x];
+        float mu2 = row_m2[x];
+        float mu11 = mu1 * mu1;
+        float mu22 = mu2 * mu2;
+        float mu12 = mu1 * mu2;
+        float nom_m = 2 * mu12 + kC1;
+        float nom_s = 2 * (row_s12[x] - mu12) + kC2;
+        float denom_m = mu11 + mu22 + kC1;
+        float denom_s = (row_s11[x] - mu11) + (row_s22[x] - mu22) + kC2;
+        row_out[x] = 1.f - ((nom_m * nom_s) / (denom_m * denom_s));
+        double d2 = row_out[x];
+        sum1[0] += d2;
+        d2 *= d2;
+        sum1[1] += d2;
+        d2 *= d2;
+        sum1[2] += d2;
+        d2 *= d2;
+        sum1[3] += d2;
+      }
+    }
+    for (int i = 0; i < 4; ++i) {
+      double e = pow(onePerPixels * sum1[i], 1.0 / (1.0 * (1 << i)));
+      plane_averages[c * 4 + i] = e;
+    }
+  }
+}
+
+}  // namespace
+
+double Msssim::Score() const {
+  double ssim = 0.0;
+  const double weight[72 * 2 + 2] = {
+      0.021973116402930787,  -11.523616782747345,    -147.37059822708056,
```

Comment:
could the optimization include some regularization term for these weights,
perhaps all of them should have the same sign (positive?) and they could have continuity regularization in scale and n

Reply:
Negative multipliers are indeed somewhat strange, and I'm not sure if I want to keep allowing it. Intuitively, it might make sense that something like 2*(4-norm) - 1*(2-norm) as an error distance could be useful — even though more error according to 2-norm leads to a better metric score, this cannot happen without the 4-norm growing too, so overall more error will still lead to a worse metric score, even if some things get a negative weight. But I'll experiment with a more constrained version where all weights are positive.

In fact I'm going to check if we really need both a multiplier and an exponent for each of the 72 norms — perhaps just one of those two is enough.

What exactly does continuity regularization mean here, and how to ensure it?